### PR TITLE
[test] clarify rpc_net & p2p_disconnect_ban functional tests 

### PR DESCRIPTION
**Pull request by amitiuttarwar**
small improvements to clarify logic in the functional tests
1. have test logic in `rpc_net.py` match run order of the test 
2. remove `connect_nodes` calls that are redundant with the automatic test setup executed by the test framework 

Noticed when I was trying to debug a test for #19725. Small changes but imo very helpful, because they initially confused me.

### CODE REVIEW

File: test/functional/rpc_net.py [hunk 1]
```diff
@@ -69,6 +72,30 @@ def test_connection_count(self):
...
+        # check both sides of bidirectional connection between nodes
+        # the address bound to on one side will be the source address for the other node
+        assert_equal(peer_info[0][0]['addrbind'], peer_info[1][0]['addr'])
+        assert_equal(peer_info[1][0]['addrbind'], peer_info[0][0]['addr'])
+        assert_equal(peer_info[0][0]['minfeefilter'], Decimal("0.00000500"))
+        assert_equal(peer_info[1][0]['minfeefilter'], Decimal("0.00001000"))
+        # check the `servicesnames` field
+        for info in peer_info:
+            assert_net_servicesnames(int(info[0]["services"], 0x10), info[0]["servicesnames"])
+
```
Comment (ipriver) [inline test/functional/rpc_net.py:98]: Why did you move `test_getpeerinfo` function declaration 75 lines upwards if you haven't modified it? It just creates unnecessary diffs for your commit.

Comment (robot-dreams) [reply test/functional/rpc_net.py:98]: Yeah, reasonable question!

- If you review one commit at a time, it's clearer to see what's going on (there's a move-only commit followed by a small functionality change)
- The commit message explains why this function was moved: "Match test order with run order"

**Comment by robot-dreams** [comment]
Concept ACK, thanks for clarifying! After trying to trace the code, I agree with you that `addnode` should add connections in both directions.

- Would it also make sense to change the "Connect nodes both ways" on line 147 below?
- The `wait_for_helper` calls in the `connect_nodes` implementation wait until the node has received a version and a verack. However, it doesn't wait until the **peer** has received a verack. Could this introduce some sort of extremely rare race condition?

File: test/functional/rpc_net.py (hunk 1 above)
Comment (ipriver) [reply test/functional/rpc_net.py:98]: @robot-dreams Agree, I missed the second commit and its title that describes this change. :blush:

**Review by epson121 — COMMENTED** [review]
Concept ACK

As @robot-dreams mentioned, these changes could be applied in L147.

Also, similar behaviour can be seen in `tests\functional\p2p_disconnect_ban.py` [L21](https://github.com/bitcoin/bitcoin/blob/9bd90db7e97cbc3f34d6f6da8097305325af0420/test/functional/p2p_disconnect_ban.py#L21) and [L80](https://github.com/bitcoin/bitcoin/blob/9bd90db7e97cbc3f34d6f6da8097305325af0420/test/functional/p2p_disconnect_ban.py#L80), e.g.

```
    def run_test(self):
        self.log.info("Connect nodes both way")
        connect_nodes(self.nodes[0], 1)
        connect_nodes(self.nodes[1], 0)
        ...
```
so updating code there as well could be an option.

As a reference, original change was introduced in this PR: https://github.com/bitcoin/bitcoin/pull/16898/files

**Comment by amitiuttarwar** [comment]
thanks for the reviews @ipriver, @robot-dreams & @epson121 ! 
updated to improve another "connect nodes both ways" instance in `p2p_disconnect_ban.py`
PR is ready for review! 

thoughts around review suggestions / questions :

> Would it also make sense to change the "Connect nodes both ways" on line 147 below?

good suggestion but I've looked into it & have found that those connections are important to leave in. Here is why: at the start of the test, the test framework creates the initial node connection behind the scenes, so we only need one additional `connect_nodes` call to get a two way connection. However in `test_getnetworkinfo` it sets the `networkactive` flag to false, which disconnects everything. So, we need to explicitly `connect_nodes` both ways in order to restore the two way connection between nodes 0 & 1. 

Let me know if this doesn't make sense / if there are further questions. You can test it out by commenting out either/both of the lines & seeing the following assertions fail. 

@robot-dreams 
> The wait_for_helper calls in the connect_nodes implementation wait until the node has received a version and a verack. However, it doesn't wait until the peer has received a verack. Could this introduce some sort of extremely rare race condition?

hm, good observation. let's think this through...
- as one data point, this function is used a lot (every time we set up a test with more than one node) so I'd imagine we'd see some sporadic failures in misc tests if this was the case. that said, we do have some spurious test failures on CI, so insufficient to rule out entirely.
- I'm trying to think of what kind of problems this could cause. message processing in bitcoin core is single threaded, so if the peer isn't processing the `VERACK` we've sent, could it mean a deeper problem where all messages aren't being processed? or taking forever or something.. 
- which leads me to wonder.. what could cause an issue?.. latency should be low, esp since the nodes are local. what are other elements that could cause long time lags?

I think overall, if a test wanted to do something like "send another message, check that message was received", it would want to use a `wait_until` for that assertion to be reliable, which should allow enough flexibility in any delay processing the `VERACK`. What do you think?

@epson121 
> Also, similar behaviour can be seen in tests\functional\p2p_disconnect_ban.py L21 and L80, e.g.

nice catch! I've updated the setup on L21. but with L80, we need to leave it in for a similar reason to what I explained earlier for the `rpc_net.py` test.

File: test/functional/rpc_net.py [hunk 2]
```diff
@@ -52,9 +52,12 @@ def set_test_params(self):
     def run_test(self):
         # Get out of IBD for the minfeefilter and getpeerinfo tests.
         self.nodes[0].generate(101)
-        # Connect nodes both ways.
+
+        # By default, the test framework sets up an addnode connection from
+        # node 1 --> node0. By connecting node0 --> node 1, this leaves us with
```
[the remark this answers is not in the document]
Comment (amitiuttarwar) [reply test/functional/rpc_net.py]: done

File: test/functional/rpc_net.py (hunk 1 above)
Comment (amitiuttarwar) [reply test/functional/rpc_net.py:98]: thanks for taking a look :)

**Comment by michaelfolkson** [comment]
Concept ACK.

Are there any edge cases where the P2P behavior of a one-sided connection is different to a two-sided connection? Do we want to test both or is a two-sided connection sufficient?

**Comment by amitiuttarwar** [comment]
@michaelfolkson not sure what you mean... I think you're asking about a subset of the tests?

**Comment by robot-dreams** [comment]
ACK 00ea3fc8e4d2f545c10970d83ef24648468deb0b

Thanks for clarifying! I previously had a brain error—for some reason I thought a single `connect_nodes` call adds connections in both directions. But that's not the case.

I tested locally to make sure that the `connect_nodes` calls you removed were indeed no-ops:

Screen Shot 2020-09-16 at 2 36 59 PM

In addition, the `connect_nodes` call you kept (on line 147) was actually doing something:

Screen Shot 2020-09-16 at 2 44 14 PM

As for my `verack` question, thanks for thinking this through! Your suggestions seem very reasonable, and in retrospect I'd be very surprised if the timing actually causes test flakiness. Although I don't fully understand what's going on here, I do think further discussion is out of scope for this PR.

**Comment by amitiuttarwar** [comment]
rebased

**Comment by amitiuttarwar** [comment]
rebased

**Comment by laanwj** [comment]
ACK 47ff5098ad5ea2c20ea387f99940a7cde6c80789
